ppl/zqd/rechandlers.go

```diff
@@ -0,0 +1,75 @@
+package zqd
+
+// Note that the handlers in this file write errors to their ResponseWriter,
+// so there are no errors returned from handle functions.
+
+// Useful CLI tests for recruiter API:
+// zqd listen -l=localhost:8020 -portfile=portfile -personality=recruiter
+// curl --header "Content-Type: application/json" -request POST --data '{"N":2}' http://localhost:8020/recruit
+// curl --header "Content-Type: application/json" -request POST --data '{"addr":"a.b.c:5000","node":"a.b"}' http://localhost:8020/register
+// curl --header "Content-Type: application/json" -request POST --data '{"addr":"a.b.c:5000"}' http://localhost:8020/unreserve
+// curl --header "Content-Type: application/json" -request POST --data '{"addr":"a.b.c:5000"}' http://localhost:8020/deregister
+// Or run system test with: make TEST=TestZq/ztests/suite/zqd/rec-curl
+
+import (
+	"net/http"
+
+	"github.com/brimsec/zq/api"
+	"github.com/brimsec/zq/zqe"
+)
+
+func handleDeregister(c *Core, w http.ResponseWriter, r *http.Request) {
+	var req api.DeregisterRequest
+	if !request(c, w, r, &req) {
+		return
+	}
+	c.workerPool.Deregister(req.Addr)
+	respond(c, w, r, http.StatusOK, api.StatusResponse{
+		Status: "ok",
+	})
+}
+
+func handleRecruit(c *Core, w http.ResponseWriter, r *http.Request) {
+	var req api.RecruitRequest
+	if !request(c, w, r, &req) {
+		return
+	}
+	ws, err := c.workerPool.Recruit(req.NumberRequested)
+	if err != nil {
+		respondError(c, w, r, zqe.ErrInvalid(err))
+		return
+	}
+	workers := make([]api.Worker, len(ws))
+	for i, e := range ws {
+		workers[i] = api.Worker{WorkerAddr: api.WorkerAddr{Addr: e.Addr}, NodeName: e.NodeName}
+	}
+	respond(c, w, r, http.StatusOK, api.RecruitResponse{
+		Workers: workers,
+	})
```

Comment:
I am wondering if these case where a request yields no workers should return an error response. This is probably something we don't want to occur regularly no? At the very least, from a dev-ops perspective if it returned an error response this would be visible in the http response log.

Reply:
That's a good point. Either the recruiter or the client should log something whenever the number returned is less than the number requested.